tokio-epoll-uring/src/util/submitting_box.rs

```diff
@@ -0,0 +1,62 @@
+//! See [`SubmittingBox`].
+
+/// A wrapper around [`Box`] with an API that forces users to spell out
+/// ownerhsip transitions of the memory between kernel and userspace.
+pub enum SubmittingBox<T>
+where
+    T: 'static,
+{
+    NotSubmitting { inner: Box<T> },
+    Submitting(*mut T),
+    Undefined,
+}
+
+unsafe impl<T> Send for SubmittingBox<T> where T: Send {}
+
+impl<T> SubmittingBox<T> {
+    pub(crate) fn new(inner: Box<T>) -> Self {
+        Self::NotSubmitting { inner }
+    }
+
+    /// [`Box::leak`] the inner box.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this function has already been called on `self`
+    /// before without a call to [`Self::ownership_back_in_userspace`] inbetween.
+    pub(crate) fn start_submitting(&mut self) -> &'static mut T {
+        match std::mem::replace(self, Self::Undefined) {
+            SubmittingBox::NotSubmitting { inner } => {
+                let leaked = Box::leak(inner);
+                *self = Self::Submitting(leaked as *mut _);
```

Comment:
`Box::into_raw` exists as well, would give you the mut directly. Had not considered `Box::leak` could be used for this. I wonder if we could run miri on this codebase.

Reply:
> However, the &mut existence at the same time is probably not an issue.

:+1: .`&mut` exists at the same time as `*mut` does. There is no problem with that, the problem is only with `&mut T `vs `&T` and `&mut T` vs `&mut T`. The `&mut T` goes out of scope later on, so from what I can tell it's safe. It's definitely not beautiful though, `Box::into_raw` is better.